lib/compat/wordpress-6.5/blocks.php

```diff
@@ -218,3 +218,38 @@ function gutenberg_process_block_bindings( $block_content, $block, $block_instan
 }
 
 add_filter( 'render_block', 'gutenberg_process_block_bindings', 20, 3 );
+
+/**
+ * Add `postId` and `postType` context to blocks that can use block bindings.
+ *
+ * @param string $args Array of arguments for registering a block type.
+ * @param string $block_name The name of the block to process.
+ * @return array The modified arguments for registering a block type.
+ */
+function gutenberg_add_context_for_block_bindings( $args, $block_name ) {
```

Comment:
And how is this done in the client, IMO, the source should also define it in the same way in both cases.

Reply:
You mean adding that context only if the source used needs it, like the case of "post meta", right? That makes sense. I'll explore that. Thanks for the feedback 🙂

Reply:
yeah, I mean the definition of the source could have something like

```
export default {
	name: 'core/post-meta',
	label: __( 'Post Meta' ),
	//...
	
	useContext: [ 'postId' ],
};
```

and same for the registration in the backend. And the framework should be wired to understand this property and register the hooks accordingly when needed.

Reply:
Perfect! That's what I understood. I agree it feels much better than what we have right now. I'll check if I can make it work.

Reply:
This is how I understood what Riad said. Every source should allow setting the required context:

```php
register_block_bindings_source( array(
    'uses_context'    => array( 'postId', 'postType' ), 
    // everything else
) );
```

Now, for all blocks that work with bindings, we would update `uses_context` to include the missing entries.

Reply:
I've been taking a look at this and I am not sure how to solve that. I've tried two different approaches:

**Adding the `uses_context` after reading the bindings attribute**

The idea was to add `uses_context` whenever a block has bindings and any attribute is connected to a source that has defined `uses_context` as in the code example.

However, the issue I'm facing is that, when the rendering of the block has started and I can access the block attributes, the registration of `uses_context` has already happened.

**Merge all the `uses_context` defined by the bindings sources**

Another approach I was considering was getting all the registered sources, merging their `uses_context` property, and adding those values to the allowed blocks.

However, we are registering sources on `init`, and it seems that runs after the registering of blocks as well.

---
@youknowriad @gziolo Any ideas how this can be approached? I am still unfamiliar with the whole rendering process.

Reply:
> However, the issue I'm facing is that, when the rendering of the block has started and I can access the block attributes, the registration of uses_context has already happened.

I expected it would be too late because the context gets exposed to the rendered block earlier during block parsing stage.

> However, we are registering sources on init, and it seems that runs after the registering of blocks as well.

I will think about it. Both block bindings sources and block types get registered on `init` so it largely depends on the order in which they get executed. I need to think about it. 

I will keep thinking about it.

Reply:
I was thinking about it and maybe we can fix that directly in core?

I was testing and it seems adding something in `WP_BLOCK` ([here](https://github.com/WordPress/wordpress-develop/blob/952004a2b34d6494fefae30bfcf4ebd069f99a2a/src/wp-includes/class-wp-block.php#L133)) could work. This is the code I used:

```php
if ( ! empty( $this->parsed_block['attrs']['metadata']['bindings'] ) ) {
	$bindings_sources = get_all_registered_block_bindings_sources();
	$this->block_type->uses_context;
	foreach ( $this->parsed_block['attrs']['metadata']['bindings'] as $attribute_name => $binding ) {
		$source_uses_context = $bindings_sources[$binding['source']]['uses_context'];
		if ( ! empty( $source_uses_context ) ) {
			$this->block_type->uses_context = array_unique(array_merge($this->block_type->uses_context, $source_uses_context));
		}
	}
}
```

Reply:
That would work best in terms of exposing the context that is required to the rendered instance of the block. I'm a little bit afraid to run this logic whenever we parse a block. It might have performance implications, so that needs to be measured accordingly.

If we were to try updating `WP_Block` as proposed then we should change this line:

https://github.com/WordPress/wordpress-develop/blob/952004a2b34d6494fefae30bfcf4ebd069f99a2a/src/wp-includes/class-wp-block.php#L134

Something like that might work

```php
$uses_context = $this->get_use_context();
if ( ! empty( $uses_context ) ) {
```

`get_uses_context` could get computed on the fly so we never update the block type.

@youknowriad, what do you think?

Reply:
I'm also not entirely certain about the impact. I had wished that we could filter the "use context" on registration but I guess "on usage" is more precise here. Let's just make sure that we do the exact same thing for both frontend and backend.

Reply:
The difference with doing it on registration is that we would updates `uses_context` even when it isn't technically necessary because we don't know at that stage whether there is going to be any attribute using the block binding. It's definitely easier to go this greedy path as we would automatically propagate changes from the server to the client.

Reply:
I don't know how the "on usage" approach would work in the editor. If the bindings are there it will work fine because the PHP code will run. But if the bindings are added later, we'd have to subscribe to the attributes changes, and run some code to add the `uses_context`.

Reply:
I've been exploring this a bit more and I think we have three possibilities:

**1. Add the `postId` to `use_context` without adding a `uses_context` in the sources**

This possibility is the current status of this pull request: [link](https://github.com/WordPress/gutenberg/pull/58554/files#diff-57852ef13af24be17ebed0abb4119a9adc146a8c3a50313d3f9a8420afda793eR231-R257).

This works for this particular case but the parameters passed to uses_context and hardcoded and sources can't define them.

We wouldn't need anything in the editor because it's already handled by the server.

**2. Add a new argument to sources and allow them defining the needed context**

I've tested this a bit more and I believe it should be possible. I'm taking that each source can pass the values like this:

```php
register_block_bindings_source(
	'core/post-meta',
	array(
		'label'              => _x( 'Post Meta', 'block bindings source' ),
		'get_value_callback' => 'gutenberg_block_bindings_post_meta_callback',
		'uses_context'       => array( 'postId', 'postType' ),
	)
);
```

From there, we can modify `wordpress-develop` [at this point](https://github.com/WordPress/wordpress-develop/blob/acad00ecabcf44a6c4e552c4fd518ee708fd131f/src/wp-includes/class-wp-block-type.php#L565) 
 (just before the `register_block_type_args` filter) to interpret that data. Something like this seems to work:

```php
if ( isset( $allowed_blocks[ $args['name'] ] ) ) {
	$sources = get_all_registered_block_bindings_sources();
	if ( ! empty( $sources ) ) {
	  foreach ( $sources as $source ) {
		  if ( $source->uses_context ) {
			  $original_use_context = isset( $args['uses_context'] ) ? $args['uses_context'] : array();
			  $args['uses_context'] = array_unique( array_merge( $original_use_context, $source->uses_context ) );
		  }
	  }
      }
}
```

This would provide more flexibility for the sources.

We wouldn't need anything in the editor because it's already handled by the server.

**3. Try to add the context "on usage".**

In the server, it could work as explained [here](https://github.com/WordPress/gutenberg/pull/58554/files#r1481141720). However, we might need some logic in the editor.

If the post has the bindings from load, there is no problem.

On the other hand, if a post has a paragraph without bindings on load, and the user adds them in the editor, it could be a problem because the context wasn't added during registration.

I am not sure if it is possible to do it "on usage" in the editor as well, or we would need to always add that context.

---

I'm inclined towards option 2, although that would mean adding the `uses_context` more than necessary if that source is not used. But I don't know how the third option would work in the editor.

Any opinions? @gziolo @youknowriad

Reply:
Option 2 sounds good to me but what about the client, I'm assuming the block will also get the useContext hooked into right?

Reply:
I was thinking about the use case where there is an existing block and I add the bindings attribute. In that case, according to my tests, it should work because the PHP registration handles that.

However, I just thought that you can start with a blank page, add an image, and then add the bindings. I assume the PHP code won't run there and the block won't have the context. For that reason, I think we have to keep (and adapt), registerBlockType filter used [here](https://github.com/WordPress/gutenberg/blob/194d208c68e4d234c43c7d0c3d3db55df56c367e/packages/block-editor/src/hooks/use-bindings-attributes.js#L117-L135).

Reply:
Random question: why we inject queryId there?

Reply:
It is not needed. I was hoping to remove it as part of this PR, and that's why I didn't include it here.

Reply:
I will go with option 2 then. I will:

* Create a PR in core to add the context defined by the sources in PHP and adapt the post meta source.
* Create a PR in Gutenberg to adapt the post meta source and the editor hook.

Reply:
Okay, I have created the first version of the potential solution:

* [Core pull request](https://github.com/WordPress/wordpress-develop/pull/6079): It adds the possibility to add the `uses_context` argument in the sources and read it before registration. It seems that code runs in the editor even if the specific block is not in the page, so we don't need the editor hook.
* I adapted the code in Gutenberg to match those changes.

I still have to review everything properly, but feel free to leave any feedback.